pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -0,0 +1,199 @@
+package triggers
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+// TODO: Register the capability in the NewDelegateFunction in chainlink/core/services/workflows.delegate.go
+
+var mercuryInfo = capabilities.MustNewCapabilityInfo(
+	"mercury-trigger",
+	capabilities.CapabilityTypeTrigger,
+	"An example mercury trigger.",
+	"v1.0.0",
+)
+
+// This Trigger Service allows for the registration and deregistration of triggers. You can also send reports to the service.
+type MercuryTriggerService struct {
+	capabilities.CapabilityInfo
+	chans                 map[workflowID]chan<- capabilities.CapabilityResponse
+	feedIdsForTriggerId   map[string][]uint64
+	mu                    sync.Mutex
+	feedIdToWorkflowId    map[uint64]string
+	triggerIdToWorkflowId map[string]string
+}
+
+type MercuryReport struct {
+	feedId               uint64
+	fullreport           []byte
+	benchmarkPrice       int64
+	observationTimestamp uint32
+}
+
+type MercuryTriggerEvent struct {
+	triggerType string // "mercury"
+	id          string // "sha256 of payload.feedId + payload.timestamp"
+	timestamp   string // "current time"
+	payload     []MercuryReport
+}
+
+var _ capabilities.TriggerCapability = (*MercuryTriggerService)(nil)
+
+func NewMercuryTriggerService() *MercuryTriggerService {
+	return &MercuryTriggerService{
+		CapabilityInfo:        mercuryInfo,
+		chans:                 map[workflowID]chan<- capabilities.CapabilityResponse{},
+		feedIdsForTriggerId:   make(map[string][]uint64),
+		triggerIdToWorkflowId: make(map[string]string),
+	}
+}
+
+// maybe use mercury-pipline reports.go report struct instead of MercuryReport struct, and then we can use the pipeline report generators
+func (o *MercuryTriggerService) ProcessReport(reports []MercuryReport) error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	currentTime := time.Now()
+	unixTimestampMillis := currentTime.UnixNano() / int64(time.Millisecond)
+	triggerIdsForReports := make(map[string][]int)
+	reportIndex := 0
+
+	for _, report := range reports {
+		// for each feed id, we need to find the triggerId associated with it.
+		reportFeedId := report.feedId
+		for triggerId, feedIds := range o.feedIdsForTriggerId {
```

Comment:
Did you consider a reverse mapping to speed up this lookup? We expect ProcessReport() calls to be much more frequent than RegisterTrigger() ones.